File: backend/app/schemas/video.py
import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict

from app.models.video import VideoStatus


class SubmitVideoRequest(BaseModel):
    url: str


class TranscriptSegmentOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    segment_index: int
    speaker_label: str
    start_seconds: float
    end_seconds: float
    text: str


class VideoSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    youtube_video_id: str
    title: str | None
    channel_name: str | None
    duration_seconds: int | None
    thumbnail_url: str | None
    status: VideoStatus
    summary: str | None = None
    created_at: datetime


class VideoDetail(VideoSummary):
    transcript_text: str | None
    error_message: str | None
    processing_started_at: datetime | None
    processing_completed_at: datetime | None
    segments: list[TranscriptSegmentOut]


class VideoStatusOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    status: VideoStatus
    error_message: str | None
    updated_at: datetime


class VideoListResponse(BaseModel):
    items: list[VideoSummary]
    total: int
    page: int
    page_size: int

File: frontend/src/types/video.ts
export type VideoStatus =
  | "queued"
  | "downloading"
  | "transcribing"
  | "summarizing"
  | "completed"
  | "failed";

export interface VideoSummary {
  id: string;
  youtube_video_id: string;
  title: string | null;
  channel_name: string | null;
  duration_seconds: number | null;
  thumbnail_url: string | null;
  status: VideoStatus;
  summary: string | null;
  created_at: string;
}

export interface TranscriptSegment {
  segment_index: number;
  speaker_label: string;
  start_seconds: number;
  end_seconds: number;
  text: string;
}

export interface VideoDetail extends VideoSummary {
  transcript_text: string | null;
  error_message: string | null;
  processing_started_at: string | null;
  processing_completed_at: string | null;
  segments: TranscriptSegment[];
}

export interface VideoStatusOut {
  id: string;
  status: VideoStatus;
  error_message: string | null;
  updated_at: string;
}

export interface VideoListResponse {
  items: VideoSummary[];
  total: number;
  page: number;
  page_size: number;
}

export const TERMINAL_STATUSES: VideoStatus[] = ["completed", "failed"];
